### ISSUE TITLE
Using leveled with multiple processes

### ISSUE THREAD

**Issue by martinsumner** [martinsumner/leveled#161]
Question has been asked about using leveled with multiple processes.  This issue is just to clarify the answer on this.

Fundamentally, leveled has been designed to have a single actor (the bookie gen_server) fronting all requests.  There are multiple processes behind that single actor (the Penciller, the Inker, the Clerks, and a separate process for each file), so that background work does not lock the Bookie - but each request to the database must sit in a single queue until the bookie is free to serve it.

If there is intended to be a long-running request (e.g. to support a fold), a read-only point-in-time snapshot can be taken to support that request offline of the main bookie (so other requests only have to wait on the time to make the snapshot, not run the fold).  The snapshots are point-in-time though, so won't be updated as new PUT requests are received by the Bookie.

There exists the potential for read requests to be managed in parallel to write requests - which would give more throughput in an environment with many CPU cores.  However, primary goal of the leveled project is for it to be used as a Riak backend, where there will always be multiple leveled instances on the machine - i.e. concurrency to make full use of the capacity of the machine is expected to be handled at a Riak level.

It would be possible, without too many code changes, to have a read-only PID to provide parallel access to leveled.  A process could be started with access to the leveled_bookie's ETS table, and direct access to the Pencller and Inker PIDs.  However, working on this is not a priority at present.  As there still needs to be a single Penciller, without significant rework, this would only provide limited parallelisation for reads.  

Leveled, depending on object and database size, can still potentially support thousands of ops per second.   If the limitation of the single fronting PID (the bookie) is a constraint, then either splitting requests over multiple leveled instances (and of course Riak can do that for you), or an alternative database would probably be better options than trying to modify leveled for greater parallelisation.

**Comment by alexdovzhanyn** [comment]
Amazing explanation, thank you!
